### PR TITLE
[Send funds] Review screen

### PR DESCRIPTION
**Pull request by elgatovital**
Handles #1785

Changes proposed in this pull request:
- Add review screen
- Tracking

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    fun onAdvancedParamsEdit() {
+        safeLaunch {
+            updateState {
+                SendAssetReviewState(
+                    viewAction = ViewAction.NavigateTo(
+                        SendAssetReviewFragmentDirections.actionSendAssetReviewFragmentToEditAdvancedParamsFragment(
+                            activeSafe.chain,
+                            safeNonce.toString(),
+                            minSafeNonce.toString(),
+                            safeTxGas.toString(),
```
Comment (biafra23) [inline app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt:72]: null is converted to "null" here and crashes in EditAdvancedParamsFragment with a numberFormatException. Maybe convert null to "0"?

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt:72]: what exactly was null here? safeNonce and minSafeNonce are required, but safeGas params can be null and EditAdvancedParamsFragment has null checks like

```
private val safeTxGas by lazy { navArgs.safeTxGas?.let { BigInteger(it) } }
```

Do you have steps to reproduce your crash?

Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt:72]: Same for `safeNonce` and `proposedSafeTxGas`

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt [hunk 2]
```diff
@@ -85,30 +92,48 @@ class SendAssetFragment : BaseViewBindingFragment<FragmentSendAssetBinding>() {
...
+            recipientInput?.asEthereumAddress()?.let {
+                updateAddress(it)
+            }
             balanceValue.text = "${selectedAsset.balanceFormatted} ${selectedAsset.symbol}"
             assetSendAmount.setAssetLogo(selectedAsset.logoUri)
             assetSendAmount.doOnTextChanged { text, _, _, _ ->
-                amountInput = nullOnThrow { BigDecimal(text.toString()) }
-                reviewButton.isEnabled = viewModel.validateInputs(recipientInput, amountInput)
+                val textString = text.toString()
+                if (textString.startsWith("0") && textString.length > 1 && textString[1] != decimalSeparator) {
```
Comment (DmitryBespalov) [inline app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:108]: is it possible to input 012345?

Comment (DmitryBespalov) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:108]: What about 0000.0001

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt (hunk 1 above)
Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt:72]: The following would work. But maybe there is a better way to ennsure they are not null here:

```
SendAssetReviewFragmentDirections.actionSendAssetReviewFragmentToEditAdvancedParamsFragment(
                            activeSafe.chain,
                            safeNonce?.let { safeNonce.toString() } ?: "0",
                            minSafeNonce.toString(),
                            safeTxGas?.let { safeTxGas.toString() } ?:  "0",
                            proposedSafeTxGas?.let { proposedSafeTxGas.toString()} ?: "0"
                        )
```

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt [hunk 3]
```diff
@@ -127,10 +152,57 @@ class SendAssetFragment : BaseViewBindingFragment<FragmentSendAssetBinding>() {
...
+                binding.assetSendAmount.setError(getString(R.string.coins_asset_send_error_amount_zero))
+            }
+
+            amountInput!! > selectedAsset.balance -> {
+                canBeReviewed = false
+                binding.assetSendAmount.setError(getString(R.string.coins_asset_send_error_amount_too_big))
+            }
+
+            amountInput!!.toPlainString().split(".")
+                .getOrNull(1)?.length ?: 0 > selectedAsset.decimals -> {
```
Comment (DmitryBespalov) [inline app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:179]: is it guaranteed to have only 1 dot?

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt (hunk 2 above)
Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:108]: no

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt (hunk 3 above)
Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:179]: yes

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt (hunk 1 above)
Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt:72]: `it` is  a String there. The String  "null"

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -159,10 +160,18 @@
     <string name="coins_asset_send">Send %1$s</string>
     <string name="coins_asset_balance">Balance:</string>
     <string name="coins_asset_send_recepient">Recipient\'s address</string>
+    <string name="coins_asset_send_from">From</string>
+    <string name="coins_asset_send_to">To</string>
     <string name="coins_asset_send_amount">Amount</string>
     <string name="coins_asset_send_max">Send max</string>
     <string name="coins_asset_send_review">Review</string>
+    <string name="coins_asset_send_review_hint">You’re about to initiate a Safe transaction. It won’t be submitted to blockchain yet.</string>
```
Comment (biafra23) [inline app/src/main/res/values/strings.xml:168]: > You’re about to initiate a Safe transaction. It won’t be submitted to blockchain yet.

I would rephrase this to:
You’re about to initiate a Safe transaction. It won’t be submitted to the blockchain yet.
or
You’re about to initiate a Safe transaction. It won’t be submitted to a blockchain yet.

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt (hunk 1 above)
Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetReviewViewModel.kt]: solved 👍

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (biafra23) [reply app/src/main/res/values/strings.xml:168]: But this is not a blocker since I have seen this error in several other places.

File: app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt (hunk 2 above)
Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:108]: no, no, and no

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:108]: we don't allow such vulgar things

Comment (DmitryBespalov) [reply app/src/main/java/io/gnosis/safe/ui/safe/send_funds/SendAssetFragment.kt:108]: perfecto, maestro! 👨‍🎨

**Review by DmitryBespalov — APPROVED** [review]

**Review by biafra23 — APPROVED** [review]
I would still check `safeNonce` not to be null. It is currently impossible but the type is nullable...

`safeNonce?.toString() ?: "0"`

**Comment by elgatovital** [comment]
> I would still check `safeNonce` not to be null. It is currently impossible but the type is nullable...
> 
> `safeNonce?.toString() ?: "0"`

added nullability check
